File: expreval/expreval.c
/* '' *********** TEST CODE YOU CAN IGNORE THIS ********* ''' */
#include <stdio.h>
#include <string.h>
#include <ctype.h>
#include <stdlib.h>

const char ops[] = { '-', '+', '/', '*', '^' };

char*
analyze( char [], int );

int
isoperator( char );

int main( void ) {

  char* expr = "-22+22";
  char* res = analyze( expr, 5 );
  printf( "\n\nInput:: %s\n", expr );
  printf( "Result:: %s\n\n", res );
  return 0;
}

int
isoperator( char c ) {
  int is = 0;
  for ( int i = 0; i < 5; i++ ) {
    if ( c == ops[i] ) {
      is = 1;
      break;
    }
  }
  return is;
}

char*
analyze( char expr[], int size ) {
  //size_t size = sizeof( expr );
  char* res = malloc( sizeof( char ) * 100 );
  char* tmp = malloc( sizeof( char ) * 1 );
  for ( int i = 0; i < size; i++ ) {
    if ( expr[i] == '-' ) {
      if ( i == 0 ) {
        strcat( res, "-" );
        for ( int j = i + 1; j < size; j++, i++ ) {
          if ( isdigit( expr[i] ) ) {
            *tmp = expr[j];
            strcat( res, tmp );
          }
          else {
            *tmp = expr[j];
            strcat( res, " " );
            strcat( res, tmp );
            strcat( res, " " );
          }
        }
      }
      else {
        strcat( res, "-" );
        strcat( res, " " );
      }
    }
    else if ( isdigit( expr[i] ) ) {
      *tmp = expr[i];
      strcat( res, tmp );
      for ( int j = i + 1; j < size; j++, i++ ) {
        if ( isdigit( expr[j] ) ) {
          *tmp = expr[j];
          strcat( res, tmp );
        }
        else {
          *tmp = expr[j];
          strcat( res, tmp );
          strcat( res, " " );
          break;
        }
      }
    }
    else if ( expr[i] == ' ' ) {
      strcat( res, " " );
    }
    else if ( isoperator( expr[i] ) ) {
      *tmp = expr[i];
      strcat( res, tmp );
      strcat( res, " " );
    }
  }
  free( tmp );
  return res;
}

File: expreval/expreval.py
''' *********** TEST CODE YOU CAN IGNORE THIS ********* '''

import re

def analyze( expr ):
    strbuild: int = ''

    for i in range( len( expr ) ):
        if expr[i] == '-':
            if i == 0:
                strbuild = strbuild + '-'
                # for j in range( i + 1, len( expr ) ):
                #     if expr[j].isdigit():
                #         strbuild = strbuild + expr[j]
                #     else:
                #         strbuild = strbuild + ' '
                #         i = i + j
                #         break
        elif expr[i].isdigit():
            strbuild = strbuild + expr[i]
            count: int = 0
            for j in range( i + 1, len( expr ) ):
                if expr[j].isdigit():
                    strbuild = strbuild + expr[j]
                else:
                    count = j
                    strbuild = strbuild + ' '
                    break
                i = i + count
        elif( expr[i] == ' '):
            strbuild = strbuild + ' '

        elif ( isOperator( expr[i] ) ):
            strbuild = strbuild + expr[i] + ' '
    return strbuild

def isOperator( ch: str ):
    ops = [ '+', '-', '/', '*', '^' ]
    # return re.match( "([*+-\/\^])", ch )
    return ch in ops

if __name__ == "__main__":
    expr = '-22457+22*2'
    res = analyze( expr )
    print( "Input:: ", expr )
    print( "Result:: ", res )
